src/Polysemy/Final.hs

```diff
@@ -208,77 +169,86 @@ liftS m = do
 -- The stateful environment will be the same as the one that the target monad
 -- is initially run in.
 -- Use 'bindS'  if you'd prefer to explicitly manage your stateful environment.
-runS :: Monad m => n a -> Sem (WithStrategy m f n) (m (f a))
-runS = fmap runM . runT
+runS :: n a -> Sem (WithStrategy m f n) (m (f a))
+runS na = bindS (const na) <*> getInitialStateS
 {-# INLINE runS #-}
 
 ------------------------------------------------------------------------------
 -- | Embed a kleisli action into the stateful environment, in terms of the final
 -- monad. You can use 'bindS' to get an effect parameter of the form @a -> n b@
 -- into something that can be used after calling 'runS' on an effect parameter
 -- @n a@.
-bindS :: Monad m => (a -> n b) -> Sem (WithStrategy m f n) (f a -> m (f b))
-bindS = fmap (runM .) . bindT
+bindS :: (a -> n b) -> Sem (WithStrategy m f n) (f a -> m (f b))
+bindS = send . HoistInterpretation
 {-# INLINE bindS #-}
 
 ------------------------------------------------------------------------------
--- | Internal function to process Strategies in terms of 'withWeaving'.
-runStrategy :: Functor f
-            => f ()
-            -> (forall x. f (n x) -> m (f x))
-            -> (forall x. f x -> Maybe x)
-            -> Sem (WithStrategy m f n) a
-            -> a
-runStrategy s wv ins (Sem m) = runIdentity $ m $ \u -> case extract u of
-  Weaving e s' _ ex' _ -> Identity $ ex' $ (<$ s') $ case e of
-    GetInitialState -> s
-    HoistInterpretation na -> embed . wv . fmap na
-    GetInspector -> Inspector ins
-{-# INLINE runStrategy #-}
-
-------------------------------------------------------------------------------
--- | Lower a 'Sem' containing only a lifted, final monad into that monad.
--- The appearance of 'Lift' as the final effect
--- is to allow the use of operations that rely on a @'LastMember' ('Lift' m)@
--- constraint.
-runFinal :: Monad m => Sem '[Final m, Embed m] a -> m a
-runFinal = usingSem $ \u -> case decomp u of
-  Right (Weaving (WithWeaving wav) s wv ex ins) ->
+-- | Lower a 'Sem' containing only a single lifted, final 'Monad' into that
+-- monad.
+runFinal :: Monad m => Sem '[Final m] a -> m a
```

Comment:
I removed the `Embed` here, since `LastMember` doesn't exist anymore, and you can use `embedToFinal` instead of processing both effects simultaneously.

That said, given that most uses of `runFinal` will be paired with a `embedToFinal` because of this, perhaps `runFinal` should just eliminate an `Embed` effect too while it's at it. It's probably more efficient that way too. Your thoughts?

Reply:
Personally I prefer having `runFinal` only interpret `Final`.

Reply:
I agree

Reply:
Then this is how `runFinal` will look.